File: problem4.js
/*

A palindromic number reads the same both ways. The largest palindrome made from
the product of two 2-digit numbers is 9009 = 91 × 99.

Find the largest palindrome made from the product of two 3-digit numbers.

*/

var palindromeNumber = function() {
  var largest = 0;
  var isPalindrome = function(num){
    var pal =  parseInt(("" + num).split("").reverse().join(""));
    if(pal === num) { return true; }
    else{ return false; }
  };

  for(var i = 999; i > 999 * 0.75; i--){
    for(var j = 999; j >= i; j--) {
      var result = i * j;
      if(isPalindrome(result) && result > largest) { largest = result; }
    }
  }

  return largest;
};

console.log(palindromeNumber());

File: erlang/problem4.erl
% A palindromic number reads the same both ways. The largest palindrome made from the product of two 2-digit numbers is 9009 = 91 × 99.

% Find the largest palindrome made from the product of two 3-digit numbers.

-module(problem4).
-export([palindrome/0]).

palindrome() ->
  run_down(999, 999).

run_down(A, B) ->
  Test = A * B,
  case is_palindrome(Test) of
    true ->
      Test;
    _ when B > (A * 0.75) ->
      run_down(A, B - 1);
    _ ->
      run_down(A - 1, A)
  end.

is_palindrome(N) ->
  Pal = list_to_integer(lists:reverse(integer_to_list(N))),
  Pal =:= N.
